File: dart/packages/verdict_rules/test/rule_test.dart
import 'package:test/test.dart';
import 'package:verdict_rules/verdict_rules.dart';

import 'test_helpers.dart';

/// A strict, 1:1 port of Python's `test_rule.py` -- FunctionRule, AndRule,
/// OrRule. Every test here has an exact Python counterpart. Dart-specific
/// coverage (structural typing for function types, an explicit `Rule`
/// implementation, the generic context migration) lives in
/// `dart_idioms_test.dart`/`generics_test.dart` instead, so this file stays
/// auditable against Python's own suite test-for-test.
void main() {
  group('FunctionRule', () {
    // Mirrors test_wraps_a_passing_predicate.
    test('wraps a passing predicate', () async {
      final rule = pass('r1');
      final result = await rule.evaluate({});
      expect(result.passed, isTrue);
      expect(result.ruleName, 'r1');
    });

    // Mirrors test_wraps_a_failing_predicate.
    test('wraps a failing predicate', () async {
      final rule = failing('r1', detail: 'nope');
      final result = await rule.evaluate({});
      expect(result.passed, isFalse);
      expect(result.detail, 'nope');
    });

    // Mirrors test_predicate_receives_the_context.
    test('predicate receives the context', () async {
      Context? seen;
      final rule = FunctionRule<Context>('r1', (ctx) async {
        seen = ctx;
        return RuleResult(ruleName: 'r1', passed: true);
      });
      await rule.evaluate({'user_id': 42});
      expect(seen, {'user_id': 42});
    });

    // Mirrors test_carries_name_and_group.
    test('carries name and group', () {
      final rule = pass('r1', group: 'g1');
      expect(rule.name, 'r1');
      expect(rule.group, 'g1');
    });

    // Mirrors test_group_defaults_to_none.
    test('group defaults to null', () {
      final rule = pass('r1');
      expect(rule.group, isNull);
    });
  });

  group('AndRule', () {
    // Mirrors test_all_pass_yields_pass.
    test('all pass yields pass', () async {
      final rule = AndRule<Context>('and1', [pass('a'), pass('b')]);
      final result = await rule.evaluate({});
      expect(result.passed, isTrue);
      expect(result.ruleName, 'and1');
    });

    // Mirrors test_one_failure_yields_fail.
    test('one failure yields fail', () async {
      final rule =
          AndRule<Context>('and1', [pass('a'), failing('b', detail: 'bad')]);
      final result = await rule.evaluate({});
      expect(result.passed, isFalse);
      expect(result.detail, contains('b'));
      expect(result.detail, contains('bad'));
    });

    // Mirrors test_short_circuits_after_first_failure.
    test('short-circuits after first failure', () async {
      final log = <String>[];
      final rule = AndRule<Context>('and1', [
        failing('a'),
        FunctionRule<Context>('c', (ctx) async {
          log.add('c');
          return RuleResult(ruleName: 'c', passed: true);
        }),
      ]);
      await rule.evaluate({});
      expect(log, isEmpty); // never reached -- 'a' already failed
    });

    // Mirrors test_data_carries_sub_results_up_to_failure.
    test('data carries sub-results up to failure', () async {
      final rule =
          AndRule<Context>('and1', [pass('a'), failing('b'), pass('c')]);
      final result = await rule.evaluate({});
      final data = result.data! as List<RuleResult>;
      expect(data.map((r) => r.ruleName), ['a', 'b']);
    });

    // Mirrors test_empty_rule_list_vacuously_passes.
    test('empty rule list vacuously passes', () async {
      final result = await AndRule<Context>('and1', []).evaluate({});
      expect(result.passed, isTrue);
    });
  });

  group('OrRule', () {
    // Mirrors test_any_pass_yields_pass.
    test('any pass yields pass', () async {
      final rule = OrRule<Context>('or1', [failing('a'), pass('b')]);
      final result = await rule.evaluate({});
      expect(result.passed, isTrue);
    });

    // Mirrors test_all_fail_yields_fail.
    test('all fail yields fail', () async {
      final rule = OrRule<Context>('or1', [failing('a'), failing('b')]);
      final result = await rule.evaluate({});
      expect(result.passed, isFalse);
      expect(result.detail, 'no sub-rule passed');
    });

    // Mirrors test_short_circuits_after_first_pass.
    test('short-circuits after first pass', () async {
      final log = <String>[];
      final rule = OrRule<Context>('or1', [
        pass('a'),
        FunctionRule<Context>('c', (ctx) async {
          log.add('c');
          return RuleResult(ruleName: 'c', passed: false);
        }),
      ]);
      await rule.evaluate({});
      expect(log, isEmpty); // never reached -- 'a' already passed
    });

    // Mirrors test_empty_rule_list_vacuously_fails.
    test('empty rule list vacuously fails', () async {
      final result = await OrRule<Context>('or1', []).evaluate({});
      expect(result.passed, isFalse);
    });
  });

  // Mirrors Python's TestExceptionPropagation in test_rule.py -- AndRule/
  // OrRule catch nothing either; a sub-rule's own exception propagates
  // straight out of evaluate(). See docs/extending/isolating-flaky-predicates/
  // for the wrapper a consumer opts into if the opposite is wanted.
  group('rule-level exception propagation', () {
    // Mirrors test_and_rule_does_not_catch_a_sub_rule_s_exception.
    test("AndRule does not catch a sub-rule's exception", () async {
      final flaky = FunctionRule<Context>('flaky', (ctx) async {
        throw StateError('boom');
      });
      expect(
        () => AndRule<Context>('and1', [pass('a'), flaky]).evaluate({}),
        throwsStateError,
      );
    });

    // Mirrors test_or_rule_does_not_catch_a_sub_rule_s_exception.
    test("OrRule does not catch a sub-rule's exception", () async {
      final flaky = FunctionRule<Context>('flaky', (ctx) async {
        throw StateError('boom');
      });
      expect(
        () => OrRule<Context>('or1', [failing('a'), flaky]).evaluate({}),
        throwsStateError,
      );
    });
  });
}

File: js/packages/verdict-rules/test/rule.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";

import { AndRule, FunctionRule, OrRule } from "../dist/index.js";
import { fail, pass } from "./helpers.js";

/**
 * A strict, 1:1 port of Python's `test_rule.py` — FunctionRule, AndRule,
 * OrRule. Every test here has an exact Python counterpart. JS-specific
 * coverage (structural typing on a bare object literal) lives in
 * `js-idioms.test.js` instead, so this file stays auditable against
 * Python's own suite test-for-test.
 */

describe("FunctionRule", () => {
  // Mirrors test_wraps_a_passing_predicate.
  it("wraps a passing predicate", async () => {
    const rule = pass("r1");
    const result = await rule.evaluate({});
    assert.equal(result.passed, true);
    assert.equal(result.ruleName, "r1");
  });

  // Mirrors test_wraps_a_failing_predicate.
  it("wraps a failing predicate", async () => {
    const rule = fail("r1", undefined, "nope");
    const result = await rule.evaluate({});
    assert.equal(result.passed, false);
    assert.equal(result.detail, "nope");
  });

  // Mirrors test_predicate_receives_the_context.
  it("predicate receives the context", async () => {
    let seen;
    const rule = new FunctionRule("r1", async (ctx) => {
      seen = ctx;
      return { ruleName: "r1", passed: true };
    });
    await rule.evaluate({ userId: 42 });
    assert.deepEqual(seen, { userId: 42 });
  });

  // Mirrors test_carries_name_and_group.
  it("carries name and group", () => {
    const rule = pass("r1", "g1");
    assert.equal(rule.name, "r1");
    assert.equal(rule.group, "g1");
  });

  // Mirrors test_group_defaults_to_none.
  it("group defaults to undefined", () => {
    const rule = pass("r1");
    assert.equal(rule.group, undefined);
  });
});

describe("AndRule", () => {
  // Mirrors test_all_pass_yields_pass.
  it("all pass yields pass", async () => {
    const rule = new AndRule("and1", [pass("a"), pass("b")]);
    const result = await rule.evaluate({});
    assert.equal(result.passed, true);
    assert.equal(result.ruleName, "and1");
  });

  // Mirrors test_one_failure_yields_fail.
  it("one failure yields fail", async () => {
    const rule = new AndRule("and1", [pass("a"), fail("b", undefined, "bad")]);
    const result = await rule.evaluate({});
    assert.equal(result.passed, false);
    assert.match(result.detail, /b/);
    assert.match(result.detail, /bad/);
  });

  // Mirrors test_short_circuits_after_first_failure.
  it("short-circuits after first failure", async () => {
    const log = [];
    const rule = new AndRule("and1", [
      fail("a"),
      new FunctionRule("c", async () => {
        log.push("c");
        return { ruleName: "c", passed: true };
      }),
    ]);
    await rule.evaluate({});
    assert.deepEqual(log, []); // never reached -- 'a' already failed
  });

  // Mirrors test_data_carries_sub_results_up_to_failure.
  it("data carries sub-results up to failure", async () => {
    const rule = new AndRule("and1", [pass("a"), fail("b"), pass("c")]);
    const result = await rule.evaluate({});
    assert.deepEqual(
      result.data.map((r) => r.ruleName),
      ["a", "b"],
    );
  });

  // Mirrors test_empty_rule_list_vacuously_passes.
  it("empty rule list vacuously passes", async () => {
    const result = await new AndRule("and1", []).evaluate({});
    assert.equal(result.passed, true);
  });
});

describe("OrRule", () => {
  // Mirrors test_any_pass_yields_pass.
  it("any pass yields pass", async () => {
    const rule = new OrRule("or1", [fail("a"), pass("b")]);
    const result = await rule.evaluate({});
    assert.equal(result.passed, true);
  });

  // Mirrors test_all_fail_yields_fail.
  it("all fail yields fail", async () => {
    const rule = new OrRule("or1", [fail("a"), fail("b")]);
    const result = await rule.evaluate({});
    assert.equal(result.passed, false);
    assert.equal(result.detail, "no sub-rule passed");
  });

  // Mirrors test_short_circuits_after_first_pass.
  it("short-circuits after first pass", async () => {
    const log = [];
    const rule = new OrRule("or1", [
      pass("a"),
      new FunctionRule("c", async () => {
        log.push("c");
        return { ruleName: "c", passed: false };
      }),
    ]);
    await rule.evaluate({});
    assert.deepEqual(log, []); // never reached -- 'a' already passed
  });

  // Mirrors test_empty_rule_list_vacuously_fails.
  it("empty rule list vacuously fails", async () => {
    const result = await new OrRule("or1", []).evaluate({});
    assert.equal(result.passed, false);
  });
});

/**
 * Mirrors Python's TestExceptionPropagation in test_rule.py -- AndRule/OrRule
 * catch nothing either; a sub-rule's own exception propagates straight out
 * of evaluate(). See docs/extending/isolating-flaky-predicates/ for the
 * wrapper a consumer opts into if the opposite is wanted.
 */
describe("rule-level exception propagation", () => {
  // Mirrors test_and_rule_does_not_catch_a_sub_rule_s_exception.
  it("AndRule does not catch a sub-rule's exception", async () => {
    const flaky = new FunctionRule("flaky", async () => {
      throw new Error("boom");
    });
    await assert.rejects(
      () => new AndRule("and1", [pass("a"), flaky]).evaluate({}),
      /boom/,
    );
  });

  // Mirrors test_or_rule_does_not_catch_a_sub_rule_s_exception.
  it("OrRule does not catch a sub-rule's exception", async () => {
    const flaky = new FunctionRule("flaky", async () => {
      throw new Error("boom");
    });
    await assert.rejects(
      () => new OrRule("or1", [fail("a"), flaky]).evaluate({}),
      /boom/,
    );
  });
});
